src/mem/metaslab.h

```diff
@@ -1,20 +1,191 @@
 #pragma once
 
-#include "../backend/metatypes.h"
+#include "../backend/backend_concept.h"
 #include "../ds/helpers.h"
 #include "../ds/seqset.h"
-#include "../mem/remoteallocator.h"
 #include "freelist.h"
 #include "sizeclasstable.h"
 
 namespace snmalloc
 {
-  // The Metaslab represent the status of a single slab.
-  class alignas(CACHELINE_SIZE) Metaslab
+  struct RemoteAllocator;
+
+  /**
+   * Base class for the templated MetaEntry.  This exists to avoid needing a
+   * template parameter to access constants that are independent of the
+   * template parameter.
+   */
+  class MetaEntryBase
   {
+    template<SNMALLOC_CONCEPT(ConceptBuddyRangeMeta) Pagemap>
+    friend class BuddyChunkRep;
+
+  protected:
+    /**
+     * Bit used to indicate this should not be considered part of the previous
+     * PAL allocation.
+     *
+     * Some platforms cannot treat different PalAllocs as a single allocation.
+     * This is true on CHERI as the combined permission might not be
+     * representable.  It is also true on Windows as you cannot Commit across
+     * multiple continuous VirtualAllocs.
+     */
+    static constexpr address_t META_BOUNDARY_BIT = 1 << 0;
+
   public:
-    MetaCommon meta_common;
+    /**
+     * This bit is set in remote_and_sizeclass to discriminate between the case
+     * that it is in use by the frontend (0) or by the backend (1).  For the
+     * former case, see mem/metaslab.h; for the latter, see backend/backend.h
+     * and backend/largebuddyrange.h.
+     *
+     * This value is statically checked by the frontend to ensure that its
+     * bit packing does not conflict; see mem/remoteallocator.h
+     */
+    static constexpr address_t REMOTE_BACKEND_MARKER = 1 << 7;
+  };
+
+  /**
+   * Remotes need to be aligned enough that the bottom bits have enough room for
+   * all the size classes, both large and small. An additional bit is required
+   * to separate backend uses.
+   */
+  static constexpr size_t REMOTE_MIN_ALIGN =
+    bits::max<size_t>(CACHELINE_SIZE, SIZECLASS_REP_SIZE) << 1;
+
+  /**
+   * The bit above the sizeclass is always zero unless this is used
+   * by the backend to represent another datastructure such as the buddy
+   * allocator entries.
+   */
+  constexpr size_t REMOTE_WITH_BACKEND_MARKER_ALIGN =
+    MetaEntryBase::REMOTE_BACKEND_MARKER;
+  static_assert(
+    (REMOTE_MIN_ALIGN >> 1) == MetaEntryBase::REMOTE_BACKEND_MARKER);
+
+  /**
+   * Entry stored in the pagemap.  See docs/AddressSpace.md for the full
+   * MetaEntry lifecycle.
+   */
+  template<typename BackendMetadata>
+  class MetaEntry : public MetaEntryBase
+  {
+    template<SNMALLOC_CONCEPT(ConceptBuddyRangeMeta) Pagemap>
+    friend class BuddyChunkRep;
```

Comment:
Isn't the need for the _frontend_ here to `friend` a _backend_ class closing the door on other backends providing new range types that use the `MetaEntry` in ways other than `BuddyChunkRep` (the large buddy range)?

Perhaps #493 was premature; it would have sufficed, I think, to ensure that `MetaEntry` and its children classes were standard layout and that the children added no new fields, making them pointer-interconvertible, and so removing any UB around casts between them.